pkg/collector/corechecks/servicediscovery/module/envs.go

```diff
@@ -153,3 +184,135 @@ func getEnvs(proc *process.Process) (map[string]string, error) {
 	}
 	return envs, nil
 }
+
+// EnvScanner reads the environment variables from /ptoc file.
+// It collects only those variables that match the target map if the map is not empty,
+// otherwise collect all environment variables.
+type EnvScanner struct {
+	file    *os.File          // open pointer to environment variables file
+	scanner *bufio.Scanner    // scanner to read file data
+	targets map[uint64]string // map of environment variables of interest
+	envs    map[string]string // collected environment variables
+}
+
+// hashBytes return hash value of a bytes array using FNV-1a hash function
+func hashBytes(b []byte) uint64 {
+	h := fnv.New64a()
+	h.Write(b)
+	return h.Sum64()
+}
+
+func buildTargetsMap() map[uint64]string {
+	targetsMap := make(map[uint64]string)
+
+	for _, target := range targetEnvs {
+		targetsMap[hashBytes([]byte(target))] = target
+	}
+	for _, target := range apm.JavaDetectorEnvs {
+		targetsMap[hashBytes([]byte(target))] = target
+	}
+	for _, target := range apm.DotNetDetectorEnvs {
+		targetsMap[hashBytes([]byte(target))] = target
+	}
+	return targetsMap
+}
+
+// nullTerminatedString used to tokenize the scanner data
+func nullTerminatedString(data []byte, atEOF bool) (advance int, token []byte, err error) {
+	if i := bytes.IndexByte(data, 0); i >= 0 {
+		return i + 1, data[0:i], nil
+	}
+	if atEOF && len(data) > 0 {
+		// end of the file, return the remaining data
+		return len(data), data, nil
+	}
+
+	return 0, nil, nil
+}
+
+// newEnvScanner returns a new [EnvScanner] to read from path.
+func newEnvScanner(proc *process.Process, onlyTargets bool) (*EnvScanner, error) {
+	envPath := kernel.HostProc(strconv.Itoa(int(proc.Pid)), "environ")
+	file, err := os.Open(envPath)
+	if err != nil {
+		return nil, err
+	}
+	var targetsMap map[uint64]string
+	if onlyTargets {
+		targetsMap = buildTargetsMap()
```

Comment:
Why are we construction this map everytime? Maybe just do somthing like https://github.com/DataDog/datadog-agent/blob/42b8cfe9f821353592ad3c444a926a5eba706a13/pkg/collector/corechecks/servicediscovery/module/impl_linux.go#L384 or something in a module init() function so it's only done once?